subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyHandlerScope.kt

```diff
@@ -43,6 +45,23 @@ private constructor(
     override val delegate: DependencyHandler
         get() = dependencies
 
+    @Deprecated(replaceWith = ReplaceWith("constraints"), message = "This method shouldn't be called because the most specific variant should be preferred by the Kotlin compiler", level = DeprecationLevel.HIDDEN)
```

Comment:
That's a creative usage of `HIDDEN`!

It does go at the heart of the problem. The wrong delegate type forced by the super method signature.

An alternative solution that might be longer term (with its own set of problems) would be to remove `DependencyHandler` from the `DependencyHandlerScope` type hierarchy altogether and keep a strict separation between API types and DSL types.